pkg/block/fetcher.go

```diff
@@ -574,4 +581,57 @@ func (f *ConsistencyDelayMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta,
 			delete(metas, id)
 		}
 	}
+
+	return nil
+}
+
+// IgnoreDeletionMarkFilter is a MetaFetcher that filters blocks that are marked for deletion; delay duration before current time.
+// The delay duration is to make sure that the replacement block can be fetched before we filter out the old block.
+type IgnoreDeletionMarkFilter struct {
+	logger          log.Logger
+	delay           time.Duration
+	bkt             objstore.BucketReader
+	deletionMarkMap map[ulid.ULID]*metadata.DeletionMark
+}
+
+// NewIgnoreDeletionMarkFilter creates IgnoreDeletionMarkFilter.
+func NewIgnoreDeletionMarkFilter(logger log.Logger, bkt objstore.BucketReader, delay time.Duration) *IgnoreDeletionMarkFilter {
+	return &IgnoreDeletionMarkFilter{
+		logger: logger,
+		bkt:    bkt,
+		delay:  delay,
+	}
+}
+
+// DeletionMarkBlocks takes in delay duration and returns block ids that were marked for deletion delay duration ago.
+func (f *IgnoreDeletionMarkFilter) DeletionMarkBlocks(delay time.Duration) []ulid.ULID {
+	blockIds := []ulid.ULID{}
+	for _, deletionMark := range f.deletionMarkMap {
+		if time.Since(time.Unix(deletionMark.DeletionTime, 0)).Seconds() > delay.Seconds() {
+			blockIds = append(blockIds, deletionMark.ID)
+		}
+	}
+	return blockIds
+}
+
+// Filter filters out blocks that are marked for deletion; delay duration before current time.
+// It also returns the blocks that can be deleted since they were uploaded delay duration before current time.
+func (f *IgnoreDeletionMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) error {
+	f.deletionMarkMap = make(map[ulid.ULID]*metadata.DeletionMark)
+
+	for id := range metas {
+		deletionMark, err := metadata.ReadDeletionMark(ctx, f.bkt, f.logger, id.String())
+		if err == metadata.ErrorDeletionMarkNotFound {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		f.deletionMarkMap[id] = deletionMark
+		if time.Since(time.Unix(deletionMark.DeletionTime, 0)).Seconds() > f.delay.Seconds() {
```

Comment:
I think there is some handy function like `timestamp.ToTime`

Reply:
Since we are mocking time now, we cannot use the function `timestamp.FromTime`. Other way would be writing our own util. Please suggest edit if something good comes to mind.

Reply:
Whatever works better (: We might write our own mockable util :+1:

Reply:
As discussed over the call, I am not mocking time in this PR